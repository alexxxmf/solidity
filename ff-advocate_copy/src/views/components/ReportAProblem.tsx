import React, { Component } from "react";
import { Image, StyleSheet, View } from "react-native";
import { Metrics } from "../../themes";
import Intercom from "react-native-intercom";
import { logger } from "../../utils/logging";
import Images from "../../themes/Images";
import Button from "./Button";

interface Props {
  userId: string;
  userEmail: string;
  userName: string;
  intercom: typeof Intercom;
}

class ReportAProblem extends Component<Props> {
  async _registerUser() {
    const userIdentifier = {
      userId: this.props.userId
    };
    logger.debug("Registering Intercom Identified User", userIdentifier);
    await this.props.intercom.registerIdentifiedUser(userIdentifier);
    const intercomAttributes = {
      email: this.props.userEmail,
      name: this.props.userName,
      custom_attributes: {
        digitalmentor: "true"
      }
    };
    logger.info("Updating Intercom attributes", {
      ...intercomAttributes,
      ...userIdentifier
    });
    await this.props.intercom.updateUser(intercomAttributes);
  }

  async _askHelp() {
    await this._registerUser();
    await this.props.intercom.displayMessenger();
  }

  async componentDidMount() {
    await this.props.intercom.registerForPush();
  }

  render() {
    return (
      <View style={styles.reportAProblem}>
        <Image
          source={Images.Problem}
          style={{
            height: Metrics.images.xlarge,
            width: Metrics.images.xlarge
          }}
          resizeMode="contain"
        />
        <Button
          testID="askHelp"
          title="Report a problem"
          type="outline"
          onPress={async () => {
            await this._askHelp();
          }}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  reportAProblem: {
    alignItems: "center"
  }
});

export default ReportAProblem;
