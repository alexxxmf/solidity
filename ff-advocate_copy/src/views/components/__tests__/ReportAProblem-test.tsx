/**
 * @jest-environment jsdom
 */
import { fireEvent, render } from "react-native-testing-library";
import ReportAProblem from "../ReportAProblem";
import React from "react";

test("ReportAProblem", async () => {
  const intercom: any = {
    registerForPush: jest.fn(),
    displayMessenger: jest.fn(),
    registerIdentifiedUser: jest.fn(),
    updateUser: jest.fn()
  };
  const renderResult = render(
    <ReportAProblem
      userId={"1234"}
      userEmail={"harry@hogwarts.com"}
      userName={"Harry Potter"}
      intercom={intercom}
    />
  );

  // Default state
  expect(renderResult.toJSON()).toMatchSnapshot("Render default");
  expect(intercom.registerForPush).toHaveBeenCalled();

  // Ask for help
  await fireEvent.press(renderResult.getByTestId("askHelp"));
  expect(renderResult.toJSON()).toMatchSnapshot("After asking help");
  expect(intercom.displayMessenger).toHaveBeenCalled();
  expect(intercom.registerIdentifiedUser.mock.calls).toMatchSnapshot(
    "Register intercom user"
  );
  expect(intercom.updateUser.mock.calls).toMatchSnapshot(
    "Update intercom user"
  );
});
