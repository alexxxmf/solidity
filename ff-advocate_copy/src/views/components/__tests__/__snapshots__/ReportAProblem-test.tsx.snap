// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportAProblem: After asking help 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../src/images/icons/Problem.png",
      }
    }
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
  <View
    style={
      Object {
        "borderRadius": 8,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="askHelp"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#111111",
            "borderRadius": 8,
            "borderWidth": 2,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#111111",
              "fontFamily": "Raleway",
              "fontSize": 16,
              "fontWeight": "bold",
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          Report a problem
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`ReportAProblem: Register intercom user 1`] = `
Array [
  Array [
    Object {
      "userId": "1234",
    },
  ],
]
`;

exports[`ReportAProblem: Render default 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../src/images/icons/Problem.png",
      }
    }
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
  <View
    style={
      Object {
        "borderRadius": 8,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="askHelp"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#111111",
            "borderRadius": 8,
            "borderWidth": 2,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#111111",
              "fontFamily": "Raleway",
              "fontSize": 16,
              "fontWeight": "bold",
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          Report a problem
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`ReportAProblem: Update intercom user 1`] = `
Array [
  Array [
    Object {
      "custom_attributes": Object {
        "digitalmentor": "true",
      },
      "email": "harry@hogwarts.com",
      "name": "Harry Potter",
    },
  ],
]
`;
